drivers: rtc: use spinlocks in RTC emul device

fgrandel opened the pull request:
Fixes: #59901

See issue description for details.

cfriedt commented:
We actually do want rtc_emul in drivers/rtc. The tests/ directory contains only unit and integration tests.

Additionally, the rtc_emul driver is able to provide the same basic service even if a platform does not have an actual hardware RTC.

This goes way back to some embedded boards where of the hw RTC was missing, it could be initialized to the **build date** for the firmware image before picking up e.g. NTP time.

fgrandel commented:
> We actually do want rtc_emul in drivers/rtc. The tests/ directory contains only unit and integration tests.

Ok, fixed.

cfriedt reviewed (dismissed):
LGTM - need @bjarki-trackunit to approve as well, ofc

Note: sorry - was trying to review each commit individually, but it really does not show up correctly via the Github UI. This is an approval only for the `[EXPERIMENTAL]` commit

cfriedt reviewed (commented):
Note: this comment refers to the `LOCKED()` changes.

I see now why you added this - actually, it's fairly useful, but the issue is that we reach into the implementation with `!__i.key` and that should not happen in the public API (`zephyr/spinlock.h`, where callers should only see `k_spin_lock()` and `k_spin_unlock()`). Additionally, the private kernel side of things _does_ need to know the implementation detail, so the same macro should probably not be used for both the public side and private side.

Personally, I would say this change is not not a direct prerequisite for updating `rtc_emul.c` to use spinlocks, so I would suggest maybe dropping it here.

cfriedt reviewed (commented):
I would just use `atomic_val_t` for both `counter` and `address`. With that, there is no need to cast.

Additionally, please fix the test to use `zassert_equal()` instead of `zassert_true()`.

Note: For the "tests: drivers: rtc: fix cast" commit

cfriedt commented on tests/drivers/rtc/rtc_api/testcase.yaml:
```diff
@@ -1,11 +1,31 @@
-      - api
-    filter: dt_alias_exists("rtc")
-    depends_on: rtc
+  drivers.rtc.rtc_api.basic: {}
+  drivers.rtc.rtc_api.alarm:
+    build_only: true
+    extra_configs:
+      - CONFIG_RTC_ALARM=y
+  drivers.rtc.rtc_api.update:
+    build_only: true
+    extra_configs:
+      - CONFIG_RTC_UPDATE=y
+  drivers.rtc.rtc_api.calibration:
+    build_only: true
+    extra_configs:
+      - CONFIG_RTC_CALIBRATION=y
+  drivers.rtc.rtc_api.full:
+    extra_configs:
+      - CONFIG_RTC_ALARM=y
+      - CONFIG_RTC_UPDATE=y
+      - CONFIG_RTC_CALIBRATION=y
```
Actually, it would be easier if all of the `=y` options could just be added to `prj.conf`. I realize that there may be unique code paths where one option is enabled and another is disabled, but this creates a fair bit of code churn for CI, and so far, all drivers support ALARM and CALIBRATE.

fgrandel replied:
Fixed.

fgrandel replied:
Reverted due to feedback by @bjarki-trackunit - see below.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -277,42 +265,41 @@ static int rtc_emul_set_time(const struct device *dev, const struct rtc_time *ti
 	}
 
-	k_mutex_lock(&data->lock, K_FOREVER);
+	K_SPINLOCK_GUARDED_SECTION(&data->lock)
```
Hmm... I see now why you added this. It's useful.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -245,27 +238,22 @@ static void rtc_emul_invoke_update_callback(const struct device *dev)
 
 #ifdef CONFIG_RTC_ALARM
-	rtc_emul_test_alarms(dev);
+		rtc_emul_test_alarms(dev);
```
If possible, please remove things that are only formatting changes.

fgrandel replied:
@cfried These are actually introduced due to the additional code block around them:

before:

```
k_mutex_lock()
rtc_emul...()
k_mutex_unlock()
```
now

```
K_SPINLOCK_GUARDED_SECTION {
      rtc_emul...()
}
```

It's just difficult to see because of the preprocessor statements.

So I believe that in this case the formatting change is inevitable (as in the other newly introduced blocks). Do you agree?

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -245,27 +238,22 @@ static void rtc_emul_invoke_update_callback(const struct device *dev)
 
 #ifdef CONFIG_RTC_UPDATE
-	rtc_emul_invoke_update_callback(dev);
+		rtc_emul_invoke_update_callback(dev);
```
Same, please remove formatting changes.

fgrandel replied:
Not ignored just resolved as a duplicate - will fix depending on the outcome of your previous comment.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -245,27 +238,22 @@ static void rtc_emul_invoke_update_callback(const struct device *dev)
-	k_mutex_lock(&data->lock, K_FOREVER);
-
-	rtc_emul_increment_tm(&data->datetime);
+	K_SPINLOCK_GUARDED_SECTION(&data->lock) {
```
I like this pattern but it would probably be better off as a separate tree-wide PR.

fgrandel replied:
Ignored for now as not being part of this PR - see discussion below.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -277,42 +265,41 @@ static int rtc_emul_set_time(const struct device *dev, const struct rtc_time *ti
-	data->datetime = (*timeptr);
-	data->datetime.tm_isdst = -1;
-	data->datetime.tm_nsec = 0;
+	K_SPINLOCK_GUARDED_SECTION(&data->lock)
```
again, great pattern but it should go into a separate tree-wide PR. For now, please use the cumbersome manual approach with `k_spin_lock()`, `k_spin_unlock()`.

fgrandel replied:
Ignored for now as not being part of this PR - see discussion below.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -277,42 +265,41 @@ static int rtc_emul_set_time(const struct device *dev, const struct rtc_time *ti
+		}
 
-		return -ENODATA;
+		(*timeptr) = data->datetime;
```
Technically no need for parenthesis, so it can just be `*timeptr`

fgrandel replied:
Fixed in a separate commit.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -277,42 +265,41 @@ static int rtc_emul_set_time(const struct device *dev, const struct rtc_time *ti
 	}
 
-	k_mutex_lock(&data->lock, K_FOREVER);
+	K_SPINLOCK_GUARDED_SECTION(&data->lock)
```
Again - great pattern. Let's fork it out to a separate tree-wide PR and do it the old-fashioned way here to narrow the focus a bit.

fgrandel replied:
Ignored for now as not being part of this PR - see discussion below.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -376,32 +362,30 @@ static int rtc_emul_alarm_get_time(const struct device *dev, uint16_t id, uint16
-	(*mask) = data->alarms[id].mask;
-
-	k_mutex_unlock(&data->lock);
+	K_SPINLOCK_GUARDED_SECTION(&data->lock)
```
Same - do it the old fashioned way, but please do either make an issue or a PR for the guarded macro.

fgrandel replied:
Ignored for now as not being part of this PR - see discussion below.

cfriedt commented on drivers/rtc/rtc_emul.c:
```diff
@@ -30,9 +23,9 @@ struct rtc_emul_data {
-	struct k_mutex lock;
+	struct k_spinlock lock;
 
-	struct rtc_emul_work_delayable dwork;
```
The mutex->spinlock conversion is agreeably a much more lightweight synchronization primitive.

Just curious - what are we saving by switching from delayable work to timer? It isn't obvious to me if there is a major benefit at this time.

fgrandel replied:
1) It's a simpler and IMO more obvious pattern (it removes some lines of code and does not need to be re-scheduled on every interrupt) - that's the main argument IMO
2) It considerably reduces the number of instructions on a time critical path (reducing latency and jitter of the generated tick signal)
3) It reduces the memory footprint of the data structure by 16 bytes IIRC (I looked at it through paholes to verify)

Do you think that justifies the change?

fgrandel replied:
Obsolete/removed - see @bjarki-trackunit 's comment below.

cfriedt reviewed (commented):
Some comments. If possible, I think it would be better to narrow the focus a bit to simply changing from mutex to spinlock. However, I do like the spinlock macro that was introduced (high-level), but it would be nice if it were a tree-wide change, where appropriate.

fgrandel commented:
> I do like the spinlock macro that was introduced (high-level), but it would be nice if it were a tree-wide change, where appropriate.

@cfriedt Did you see that this is a draft PR for that reason? The first commit introducing the macro does not belong to this PR, this PR will be rebased (or changed back to not using the macro) once #59884 has been merged or rejected. Is it ok, if I just ignore your comments re this pattern then? Or do I misunderstand you?

cfriedt commented:
> @cfriedt Did you see that this is a draft PR for that reason? The first commit introducing the macro does not belong to this PR, this PR will be rebased (or changed back to not using the macro) once #59884 has been merged or rejected. Is it ok, if I just ignore your comments re this pattern then? Or do I misunderstand you?

Yes of course.

Would you prefer I comment here or on Discord?

fgrandel commented:
> Would you prefer I comment here or on Discord?

To me both is equally fine. There seems to be some disagreement in the community depending on who you ask. Some see Discord critically because it's not searchable, others want to reduce verbosity in public issues. Choose according to your preferfence, I'm fine with any working style as long as it leads to a solution. :-)

fgrandel commented:
> I would just use `atomic_val_t` for both `counter` and `address`. With that, there is no need to cast.

@cfriedt I think there is no cast of counter any more. Probably github tricked you in looking at an old version of the commit? The only cast left is that of a pointer to atomic_val_t which is probably inevitable. Do you agree?

Commit:
tests: drivers: rtc: zassert_true -> zassert_equal

Using the more specialized zassert_equal() macro where appropriate.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
> Additionally, please fix the test to use `zassert_equal()` instead of `zassert_true()`.

Fixed.

bjarki-trackunit commented on drivers/rtc/Kconfig:
```diff
@@ -2,7 +2,8 @@
 
 menuconfig RTC
-	bool "Real-Time Clock (RTC) drivers"
+	bool "Real-Time Clock (RTC) drivers [EXPERIMENTAL]"
```
Device driver APIs are not experimental features, they follow the lifecycle described [here](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#api-lifecycle) The RTC device driver API is at this time `unstable` by the actual descriptions used for device driver APIs. This is therefore an inappropriate change.

cfriedt replied:
Yeah, I wouldn't change the spinlock API in this pr. If the api is experimental or not, it's kind of irrelevant.

fgrandel replied:
@cfriedt That spinlock stuff is now merged in main, so this PR is no longer a draft. The question I asked above wasn't so much about the spinlock (which I never intended to be part of this PR) but about marking the RTC driver API experimental. I'm a bit confused, because it is shown as experimental in the [API overview](https://docs.zephyrproject.org/latest/develop/api/overview.html) plus experimental API drivers are usually marked as such in Kconfig to avoid confusion and (more importantly) flagged as `select EXPERIMENTAL` so they produce automated warnings when used which you confirmed in your first review. @bjarki-trackunit  disagreed on that point, that's why I asked you both to quickly co-ordinate, so that I know what the final consensus will be on that point and finalize this PR then.

fgrandel replied:
Clarified with cfried that this change will be removed. The corresponding ticket was closed as "won't fix".

bjarki-trackunit commented on drivers/rtc/rtc_emul.c:
```diff
@@ -10,13 +10,6 @@
 #include <zephyr/device.h>
 #include <zephyr/drivers/rtc.h>
 
-struct rtc_emul_data;
```
This change us better achieved by using the CONTAINER_OF macro, which removes the need for this structure, while keeping the work queue item.

fgrandel replied:
Obsolete: change was reverted.

bjarki-trackunit commented on drivers/rtc/rtc_emul.c:
```diff
@@ -28,15 +21,9 @@ struct rtc_emul_alarm {
-
-	struct k_mutex lock;
-
-	struct rtc_emul_work_delayable dwork;
```
Usage of the work queue item is preferred to a timer since work, like invoking a callback, shall not be performed from within a timer callback. Work must be delegated to a thread, like the work queue. This is the reason ISR callbacks delegate work to be performed to the work queue.

Keep in mind that this driver is not time critical, it is not a real piece of hardware meant to actually be used as a functional RTC, its an emulator, which provides similar behavior to an RTC, allowing for rapid application development if real hardware is not available.

fgrandel replied:
Timer conversion is now removed from change set.

bjarki-trackunit commented on include/zephyr/spinlock.h:
```diff
@@ -29,6 +29,56 @@ extern "C" {
  * @{
  */
 
+/**
```
This change must be placed in its own PR.

fgrandel replied:
See PR description - this change already is in a separate change set. That's why the PR is a draft. Depending on the outcome of the other PR this PR will either stay as it is or be changed to use the classical spinlock pattern.

bjarki-trackunit commented on kernel/include/kernel_internal.h:
```diff
@@ -227,10 +227,10 @@ void z_mem_manage_init(void);
  */
 void z_mem_manage_boot_finish(void);
 
-#define LOCKED(lck) for (k_spinlock_key_t __i = {},			\
```
This should also be moved out of this PR

fgrandel replied:
Duplicate.

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/prj.conf:
```diff
@@ -1,6 +1,9 @@
-CONFIG_RTC=y
 CONFIG_ZTEST=y
 CONFIG_ZTEST_NEW_API=y
+
```
The RTC API only tests the mandatory RTC features by default. Additional features like the alarm might not be implemented, which will lead to failed tests for those devices. This change shall be undone for that reason.

fgrandel replied:
The change was reverted/removed from the change set.

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/src/test_update_callback.c:
```diff
@@ -18,30 +18,30 @@ static void test_rtc_update_callback_handler(const struct device *dev, void *use
 	int ret;
 	uint32_t counter;
-	uint32_t address;
+	atomic_val_t address;
```
This value is not atomic, its storing the value of an atomic. This change introduces a bug if the atomic_val_t is not a uint32_t, as you typecast it on line 54.

fgrandel replied:
The change was removed from the change set.

bjarki-trackunit commented on tests/drivers/rtc/rtc_api/testcase.yaml:
```diff
@@ -9,3 +9,5 @@ tests:
       - api
     filter: dt_alias_exists("rtc")
     depends_on: rtc
+    integration_platforms:
```
native_posix is not an integration platform, this test suite is built to run on any platform that has an RTC, real and emulated like native_posix or qemu [see](https://docs.zephyrproject.org/latest/develop/test/twister.html)

fgrandel replied:
The change was removed from the change set.

bjarki-trackunit commented on tests/drivers/rtc/rtc_api_helpers/testcase.yaml:
```diff
@@ -8,3 +8,5 @@ tests:
       - rtc
       - api
       - helpers
+    integration_platforms:
```
See previous comment

fgrandel replied:
Duplicate.

bjarki-andreasen reviewed (changes requested):
I agree with changing the rtc_emul driver to use a spinlock instead of a mutex as @cfriedt has previously noted as well.

Changing the zephyr test stuite to use zassert_equal instead of zassert_true I am apathetic to.

Intoducing a change to the spinlock API is a big change that requires its own PR, I have no issue with that change in general, but I lack the knowledge required to review that change.

As my comments reveal, the remaining changes are either incorrect or IMO unwarranted.

Commit:
drivers: rtc: rtc_emul: replace mutex with spinlock

Replaces the mutex with a lightweight spinlock.

Fixes: #59901

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
driver: rtc: rtc_emul: remove redundant brackets

Change due to review comment: remove redundant brackets for pointer
value access.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

cfriedt reviewed (dismissed):
Just a mention, maybe for @bjarki-trackunit to follow up, but `zassert_equal(x, 0, ..)` can be replaced with `zassert_ok(x, ...)`

bjarki-andreasen reviewed (changes requested):
Looks good, however, as @cfriedt discovered, the macro `zassert_ok` is safer than `zassert_equal` where applicable. I will request you use that macro instead where applicable. I will sanity check the changes locally and approve the PR once its ready :)

fgrandel commented:
> the macro `zassert_ok` is safer than `zassert_equal` 

@bjarki-trackunit  Fixed.

bjarki-andreasen reviewed (approved):
Emulated RTC driver and RTC API test suite behavior validated. Nice work :)
